galley/pkg/kube/source/source.go

```diff
@@ -49,6 +53,68 @@ func New(k kube.Interfaces, resyncPeriod time.Duration, cfg *converter.Config) (
 	return newSource(k, resyncPeriod, cfg, kube_meta.Types.All())
 }
 
+// VerifyCRDPresence verifies that all expected CRDs are registered
+// with the k8s kube-apiserver.
+func VerifyCRDPresence(k kube.Interfaces) error {
+	cs, err := k.APIExtensionsClientset()
+	if err != nil {
+		return err
+	}
+	return verifyCRDPresence(cs, kube_meta.Types.All())
+}
+
+var (
+	crdPresencePollInterval = 500 * time.Millisecond
+	crdPresensePollTimeout  = time.Minute
+)
+
+func verifyCRDPresence(cs clientset.Interface, specs []kube.ResourceSpec) error {
+	crdToFind := make(map[string]struct{})
+	for _, spec := range specs {
+		name := fmt.Sprintf("%s.%s", spec.Plural, spec.Group)
+		crdToFind[name] = struct{}{}
+	}
+
+	err := wait.Poll(crdPresencePollInterval, crdPresensePollTimeout, func() (bool, error) {
+		var errs error
+	nextCRD:
+		for name := range crdToFind {
+			crd, err := cs.ApiextensionsV1beta1().CustomResourceDefinitions().Get(name, v1.GetOptions{})
+			if err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("could not find %v: %v", name, err))
+				continue nextCRD
+			}
+
+			for _, cond := range crd.Status.Conditions {
+				switch cond.Type {
+				case apiextensionsv1beta1.Established:
+					if cond.Status == apiextensionsv1beta1.ConditionTrue {
+						log.Infof("Found CRD %q", name)
+						delete(crdToFind, name)
```

Comment:
Presumably deleting while looping over a slice is safe?

Reply:
Yes. 

(from https://golang.org/ref/spec#For_statements)
> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next. ___If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced___. If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.